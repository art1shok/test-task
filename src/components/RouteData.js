import React from 'react';
import {
  RouteContainer,
  RouteDataStyled,
  RouteDataContainer,
  RouteDescription
} from '../styled/RouteData.styled';

const handleLandingTime = (startHours, startMinutes, flightTimeHours, flightTimeMinutes) => {
  let Hours = startHours + flightTimeHours;
  let Minutes = startMinutes + flightTimeMinutes;

  while (Hours >= 24) {
    Hours -= 24;
  }

  if (Minutes > 60) {
    Minutes -= 60;
  }

  return {
    hours: Hours,
    minutes: Minutes
  };
};

export const RouteData = ({ way, date, duration, stops }) => {
  const hh = new Date(date).getHours();
  const mm = new Date(date).getMinutes();
  const flightHours = Math.floor(duration / 60);
  const flightMinutes = duration %= 60;

  const landingTime = handleLandingTime(hh, mm, flightHours, flightMinutes);

  return (
    <RouteContainer md={{ span: 4 }}>
      <RouteDataContainer>
        <RouteDescription>{way}</RouteDescription>
        <RouteDataStyled>
          {`${hh < 10 ? '0' + hh : hh}:${mm < 10 ? '0' + mm : mm} -
           ${landingTime.hours < 10 ? '0' + landingTime.hours : landingTime.hours}:${
            landingTime.minutes < 10 ? '0' + landingTime.minutes : landingTime.minutes}`}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer md={{ span: 4 }}>
        <RouteDescription>В пути</RouteDescription>
        <RouteDataStyled>
          {`${flightHours}ч ${flightMinutes}м`}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer md={{ span: 3 }}>
        <RouteDescription>
          {stops.length === 0 ? 'Без пересадок' : stops.length === 1 ? '1 пересадка' : `${stops.length} пересадки`}
        </RouteDescription>
        <RouteDataStyled>{stops.join(', ')}</RouteDataStyled>
      </RouteDataContainer>
    </RouteContainer>
  );
};
