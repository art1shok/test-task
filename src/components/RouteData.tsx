import React, { FC, useMemo } from 'react';
import {
  RouteContainer,
  RouteDataStyled,
  RouteDataContainer,
  RouteDescription,
} from '../styled/RouteData.styled';

interface Props {
  way: string;
  date: string;
  duration: number;
  stops: string[];
}

const handleLandingTime = (startHours: number, startMinutes: number,
  flightTimeHours: number, flightTimeMinutes: number) => {
  let Hours = startHours + flightTimeHours;
  let Minutes = startMinutes + flightTimeMinutes;

  while (Hours >= 24) {
    Hours -= 24;
  }

  if (Minutes > 60) {
    Minutes -= 60;
  }

  return {
    hours: Hours,
    minutes: Minutes,
  };
};

export const RouteData: FC<Props> = ({
  way, date, duration, stops,
}) => {
  const hh = new Date(date).getHours();
  const mm = new Date(date).getMinutes();
  const flightHours = Math.floor(duration / 60);
  const flightMinutes = duration % 60;

  const landingTime = handleLandingTime(hh, mm, flightHours, flightMinutes);

  const addZeroes = (time: number) => (time < 10 ? `0${time}` : time);

  const formattedTime = useMemo(() =>
    `${addZeroes(hh)}:${addZeroes(mm)} - ${addZeroes(landingTime.hours)}:${addZeroes(landingTime.minutes)}`,
  [hh, landingTime.hours, landingTime.minutes, mm]);

  const handleStops = (transfers: string[]) => {
    if (transfers.length === 0) {
      return 'Без пересадок';
    } if (transfers.length === 1) {
      return '1 пересадка';
    }
    return `${transfers.length} пересадки`;
  };

  const transfers = handleStops(stops);

  return (
    <RouteContainer>
      <RouteDataContainer>
        <RouteDescription className="way">{way}</RouteDescription>
        <RouteDataStyled className="time">
          {formattedTime}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer>
        <RouteDescription>В пути</RouteDescription>
        <RouteDataStyled>
          {`${flightHours}ч ${flightMinutes}м`}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer>
        <RouteDescription className="stops">
          {transfers}
        </RouteDescription>
        <RouteDataStyled className="transfers">{stops.join(', ')}</RouteDataStyled>
      </RouteDataContainer>
    </RouteContainer>
  );
};
