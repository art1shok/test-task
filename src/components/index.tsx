export { FlightTable } from './FlightTable';
export { Ticket } from './Ticket';
export { RouteData } from './RouteData';
