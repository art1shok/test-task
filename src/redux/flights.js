import { put, call, takeEvery } from 'redux-saga/effects';
import { apiService } from '../api/apiService';

const SET_ERROR = 'test-task/flights/SET_ERROR';
const GET_SEARCH_ID = 'test-task/flights/GET_SEARCH_ID';
const GET_TICKETS = 'test-task/flights/GET_TICKETS';

const initialState = {
  searchId: null,
  ticketsInfo: null,
  error: null
};

const reducer = (state = initialState, action) => {

  const { payload } = action;
  switch (action.type) {

    case GET_SEARCH_ID: {

      return {
        ...state,
        searchId: payload
      };
    }

    case GET_TICKETS: {

      return {
        ...state,
        ticketsInfo: payload
      };
    }

    case SET_ERROR: {

      return {
        ...state,
        error: payload,
      };
    }

    default:
      return state;
  }
};

export default reducer;

export const getSearchId = (data) => ({
  type: GET_SEARCH_ID,
  payload: data
});

export const getTickets = (data) => ({
  type: GET_TICKETS,
  payload: data
});

export const setError = (error) => ({
  type: SET_ERROR,
  payload: error,
});

export function * sagaWatcher () {
  yield takeEvery(GET_SEARCH_ID, sagaWorker);
}

export function * sagaWorker () {
  try {
    const result =
      yield call(apiService.getSearchId);
    const data =
      yield call(() => apiService.getTickets(result.searchId));
    yield put(getTickets(data));
  } catch (e) {
    yield put(setError(e));
  }
}
