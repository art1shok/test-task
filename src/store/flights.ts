import { put, call, takeEvery } from 'redux-saga/effects';
import { apiService } from '../api/apiService';
import {
  FlightsActionTypes,
  FlightsState,
  GetTicketsAction,
  SetErrorAction,
  TicketsInfo,
} from './flights.types';

export const SET_ERROR = 'test-task/flights/SET_ERROR';
export const GET_SEARCH_ID = 'test-task/flights/GET_SEARCH_ID';
export const GET_TICKETS = 'test-task/flights/GET_TICKETS';

const initialState: FlightsState = {
  ticketsInfo: {
    tickets: [],
    stop: true,
  },
  error: null,
};

const flightsReducer = (
  state = initialState,
  action: FlightsActionTypes,
): FlightsState => {
  switch (action.type) {
    case GET_SEARCH_ID: {
      return {
        ...state,
      };
    }

    case GET_TICKETS: {
      return {
        ...state,
        ticketsInfo: (action as GetTicketsAction).payload,
      };
    }

    case SET_ERROR: {
      return {
        ...state,
        error: (action as SetErrorAction).payload,
      };
    }

    default:
      return state;
  }
};

export default flightsReducer;

export const getSearchId = () => ({
  type: GET_SEARCH_ID,
});

export const getTickets = (data: TicketsInfo) => ({
  type: GET_TICKETS,
  payload: data,
});

export const setError = (error: any) => ({
  type: SET_ERROR,
  payload: error,
});

export function* sagaWatcher() {
  yield takeEvery(GET_SEARCH_ID, sagaWorker);
}

export function* sagaWorker() {
  try {
    const result = yield call(apiService.getSearchId);
    const data = yield call(() => apiService.getTickets(result.searchId));
    yield put(getTickets(data));
  } catch (error) {
    yield put(setError(error.message));
  }
}
