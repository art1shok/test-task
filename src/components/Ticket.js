import React from 'react';
import { RouteData } from './RouteData';

import {
  CardHead,
  CardLogo,
  Price,
  TicketWrapper
} from '../styled/Ticket.styled';

export const Ticket = ({ price, route }) => (
  <TicketWrapper>
    <CardHead>
      <Price>
        {`${price} ла`}
      </Price>
      <CardLogo src="../../../assets/images/logo.svg" alt="logo"/>
    </CardHead>
    {route && route.map((item, index) => (
        <RouteData
          way={`${item.origin} - ${item.destination}`}
          stops={item.stops}
          date={item.date}
          duration={item.duration}
          key={index}
        />
      )
    )}

  </TicketWrapper>
);

