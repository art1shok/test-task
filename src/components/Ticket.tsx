import React, {FC} from 'react';
import {RouteData} from './RouteData';

import {
  CardHead,
  CardLogo,
  Price,
  TicketWrapper
} from '../styled/Ticket.styled';
import {Segment} from "../store/flights.types";

interface Props {
  price: number;
  route: [
    Segment,
    Segment
  ]
}

export const Ticket: FC<Props> = ({price, route}) => (
  <TicketWrapper>
    <CardHead>
      <Price>
        {`${price.toLocaleString()} ла`}
      </Price>
      <CardLogo src="../../../assets/images/logo.svg" alt="logo"/>
    </CardHead>
    {route && route.map((item: any, index: number) => (
      <RouteData
        way={`${item.origin} - ${item.destination}`}
        stops={item.stops}
        date={item.date}
        duration={item.duration}
        key={index}
      />
    ))}
  </TicketWrapper>
);

