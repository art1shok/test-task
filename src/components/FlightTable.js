import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { FormGroup } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import { Ticket } from './Ticket';
// import { selectTicketsInfo } from '../redux/selectors';
import { getSearchId } from '../redux/flights';
import { ticket } from '../mocks/ticket.mock'

import {
  Background,
  ButtonContainer,
  CheckboxContainer,
  FilterContainer,
  Image,
  ImageContainer,
  StyledButton,
  StyledCheckbox,
  StyledContainer,
  StyledText,
  TicketsContainer,
} from '../styled/FlightTable.styled';

const filters = {
  none: 0,
  one: 1,
  two: 2,
  three: 3,
};

export const FlightTable = () => {
  const [sortType, setSortType] = useState('cheapest');
  const history = useHistory();
  const dispatch = useDispatch();
  // TODO: remove comments when API is working
  //const info = useSelector(selectTicketsInfo);
  const info = ticket;

  const paramsSet = useMemo(() => {
    const newParamsSet = {
      all: false,
      one: false,
      two: false,
      three: false,
      none: false,
    };

    const paramsArr = history.location.search !== ''
      ? history.location.search.substring(1)
        .split('&')
        .filter(item => item !== '')
      : [];

    paramsArr.forEach(item => {
      newParamsSet[item] = true;
    })
    return newParamsSet;

  }, [history.location.search]);

  useEffect(() => {
    dispatch(getSearchId());
  }, [dispatch]);

  const handleChange = useCallback((e) => {
    const { name } = e.target;
    let queryString = '?';

    Object.entries(paramsSet).forEach(([filter, isActive], index) => {
      if ((filter === name && !isActive) || (filter !== name && isActive)) {
        queryString += `${filter}&`;
      }
    });

    history.push({
      pathname: '/',
      search: queryString
    });
  }, [history, paramsSet]);

  const sortCheapest = (a, b) => a.price > b.price ? 1 : -1;

  const sortFastest = (a, b) => a.segments[0].duration > b.segments[0].duration ? 1 : -1;

  const filteredTickets = useMemo(() => {
    if (paramsSet['all'] || Object.values(paramsSet).every(item => !item)) {
      return info;
    }

    return info.filter((ticket) => {
        let isAppropriate = false;
        Object.entries(paramsSet).forEach(([filterName, isActive]) => {
          if (isActive && ticket.segments.every(segment => segment.stops.length === filters[filterName])) {
            isAppropriate = true;
          }
        });
        return isAppropriate;
      }
    );

  }, [info, paramsSet]);

  const sortedTickets = useMemo(() =>
      filteredTickets.sort(sortType === 'cheapest' ? sortCheapest : sortFastest),
    [filteredTickets, sortType]
  );


  const getVariant = (type) => sortType === type ? 'primary' : 'light';

  return (
    <Background>
      <StyledContainer>
        <ImageContainer>
          <Image src="../assets/images/plane.svg" alt="plane"/>
        </ImageContainer>

        <FilterContainer>
          <CheckboxContainer>
            <StyledText>
              Колличество пересадок
            </StyledText>
            <FormGroup controlId="formBasicCheckbox">
              <StyledCheckbox
                type="checkbox"
                label="Все"
                id="all"
                name="all"
                checked={paramsSet['all']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="Без пересадок"
                id="none"
                name="none"
                checked={paramsSet['none']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="1 пересадка"
                name="one"
                id="one"
                checked={paramsSet['one']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="2 пересадки"
                name="two"
                id="two"
                checked={paramsSet['two']}
                onChange={handleChange}
              />
              <StyledCheckbox
                type="checkbox"
                label="3 пересадки"
                name="three"
                id="three"
                checked={paramsSet['three']}
                onChange={handleChange}
              />
            </FormGroup>
          </CheckboxContainer>
          <TicketsContainer>
            <ButtonContainer size="lg">
              <StyledButton
                variant={getVariant('cheapest')}
                onClick={() => setSortType('cheapest')}
              >
                Самый дешевый
              </StyledButton>
              <StyledButton
                variant={getVariant('fastest')}
                onClick={() => setSortType('fastest')}
              >
                Самый быстрый
              </StyledButton>
            </ButtonContainer>
            {sortedTickets && sortedTickets.slice(0, 5).map((item, index) => (
              <Ticket
                key={index}
                price={item.price}
                route={item.segments}
              />
            ))}

          </TicketsContainer>
        </FilterContainer>
      </StyledContainer>
    </Background>
  );
};
