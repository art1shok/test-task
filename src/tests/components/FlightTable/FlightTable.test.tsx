import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {
  shallow, configure, ShallowWrapper, ShallowRendererProps,
} from 'enzyme';
import configureStore from 'redux-mock-store';
import * as reactRedux from 'react-redux';
import 'jest-enzyme';

import { Store } from 'redux';
import { FlightTable } from '../../../components';
import { FlightsState } from '../../../store/flights.types';
import {
  StyledButton,
  StyledCheckbox,
} from '../../../styled/FlightTable.styled';

configure({
  adapter: new Adapter(),
});

const initialState : FlightsState = {
  ticketsInfo: {
    tickets: [
      {
        price: 1120,
        carrier: 'AI',
        segments: [
          {
            origin: 'MAK',
            destination: 'MAV',
            date: '10/03/2017 07:29:46-0700',
            stops: [
              'JAK',
              'DAV',
            ],
            duration: 1200,
          },
          {
            origin: 'MAV',
            destination: 'MAK',
            date: '10/03/2017 07:29:46-0700',
            stops: [
              'DWV',
              'DAV',
            ],
            duration: 1890,
          },
        ],
      },
      {
        price: 2213,
        carrier: 'AI',
        segments: [
          {
            origin: 'MAK',
            destination: 'MAV',
            date: '10/03/2017 05:29:46-0700',
            stops: [
              'DAV',
            ],
            duration: 1109,
          },
          {
            origin: 'MAV',
            destination: 'MAK',
            date: '10/03/2017 07:29:46-0700',
            stops: [
              'DWv',
            ],
            duration: 768,
          },
        ],
      },
      {
        price: 720,
        carrier: 'AI',
        segments: [
          {
            origin: 'MAK',
            destination: 'MAV',
            date: '10/03/2017 07:29:46-0700',
            stops: [],
            duration: 220,
          },
          {
            origin: 'MAV',
            destination: 'MAK',
            date: '10/03/2017 07:29:46-0700',
            stops: [],
            duration: 1800,
          },
        ],
      },
      {
        price: 12400,
        carrier: 'AI',
        segments: [
          {
            origin: 'MAK',
            destination: 'MAV',
            date: '10/03/2017 07:29:46-0700',
            stops: [
              'ADW',
              'AWW',
              'AN',
            ],
            duration: 740,
          },
          {
            origin: 'MAV',
            destination: 'MAK',
            date: '10/03/2017 07:29:46-0700',
            stops: [
              'AW',
              'ADD',
              'DWW',
            ],
            duration: 1800,
          },
        ],
      },
    ],
    stop: false,
  },
  error: null,
};

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual<any>('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
    location: {
      search: 'string',
    },
  }),
}));

describe('App component', () => {
  let wrapper: ShallowWrapper<any>;
  let store: Store;

  beforeEach(() => {
    store = configureStore()(initialState);

    jest
      .spyOn(reactRedux, 'useSelector')
      .mockImplementation(() => store.getState().ticketsInfo);

    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => store.dispatch);
    const context = { store };
    wrapper = shallow(
      <FlightTable />,
      context as ShallowRendererProps,
    ).dive();
  });
  it('Flight Table render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Flight Table should have 5 filter checkboxes',
    () => {
      expect(wrapper.find(StyledCheckbox)).toHaveLength(5);
    });

  it('Flight Table should have 2 sorting buttons',
    () => {
      expect(wrapper.find(StyledButton)).toHaveLength(2);
    });

  it('Should have onClick function on the  button',
    () => {
      expect(wrapper.find(StyledButton).at(1).simulate('click')).toHaveLength(1);
    });

  it('Should receive event function on click the checkbox',
    () => {
      expect(wrapper.find(StyledCheckbox).at(1).simulate('event')).toHaveLength(1);
    });
});
