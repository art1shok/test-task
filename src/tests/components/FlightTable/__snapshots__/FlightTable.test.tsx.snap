// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Flight Table render correctly 1`] = `
<div
  className="FlightTablestyled__Background-sc-1vwjd7g-0 igErIY"
>
  <FlightTablestyled__StyledContainer>
    <FlightTablestyled__ImageContainer>
      <FlightTablestyled__Image
        alt="plane"
        src="../assets/images/plane.svg"
      />
    </FlightTablestyled__ImageContainer>
    <FlightTablestyled__FilterContainer>
      <FlightTablestyled__CheckboxContainer>
        <FlightTablestyled__StyledText>
          Колличество пересадок
        </FlightTablestyled__StyledText>
        <FormGroup
          controlId="formBasicCheckbox"
        >
          <FlightTablestyled__StyledCheckbox
            checked={false}
            id="all"
            label="Все"
            name="all"
            onChange={[Function]}
            type="checkbox"
          />
          <FlightTablestyled__StyledCheckbox
            checked={false}
            id="none"
            label="Без пересадок"
            name="none"
            onChange={[Function]}
            type="checkbox"
          />
          <FlightTablestyled__StyledCheckbox
            checked={false}
            id="one"
            label="1 пересадка"
            name="one"
            onChange={[Function]}
            type="checkbox"
          />
          <FlightTablestyled__StyledCheckbox
            checked={false}
            id="two"
            label="2 пересадки"
            name="two"
            onChange={[Function]}
            type="checkbox"
          />
          <FlightTablestyled__StyledCheckbox
            checked={false}
            id="three"
            label="3 пересадки"
            name="three"
            onChange={[Function]}
            type="checkbox"
          />
        </FormGroup>
      </FlightTablestyled__CheckboxContainer>
      <FlightTablestyled__TicketsContainer>
        <FlightTablestyled__ButtonContainer
          size="lg"
        >
          <FlightTablestyled__StyledButton
            onClick={[Function]}
            variant="primary"
          >
            Самый дешевый
          </FlightTablestyled__StyledButton>
          <FlightTablestyled__StyledButton
            onClick={[Function]}
            variant="light"
          >
            Самый быстрый
          </FlightTablestyled__StyledButton>
        </FlightTablestyled__ButtonContainer>
      </FlightTablestyled__TicketsContainer>
    </FlightTablestyled__FilterContainer>
  </FlightTablestyled__StyledContainer>
</div>
`;
